x/campaign/keeper/msg_server_add_mainnet_vesting_account_test.go

```diff
@@ -0,0 +1,201 @@
+				VestingOptions: sample.ShareVestingOptions(),
+			},
+			err: profiletypes.ErrCoordAddressNotFound,
+		},
+		{
+			name: "invalid coordinator id",
+			msg: types.MsgAddMainnetVestingAccount{
+				Coordinator:    coordAddr2,
+				CampaignID:     campaign1.Id,
+				Address:        addr1,
+				Shares:         sample.Shares(),
+				VestingOptions: sample.ShareVestingOptions(),
+			},
+			err: profiletypes.ErrCoordInvalid,
+		},
+		{
+			name: "campaign already in mainnet",
+			msg: types.MsgAddMainnetVestingAccount{
+				Coordinator:    coordAddr2,
+				CampaignID:     campaign2.Id,
+				Address:        addr1,
+				Shares:         sample.Shares(),
+				VestingOptions: sample.ShareVestingOptions(),
+			},
+			err: types.ErrMainnetInitialized,
+		},
+		{
+			name: "allocated shares greater them total shares",
+			msg: types.MsgAddMainnetVestingAccount{
+				Coordinator:    coordAddr3,
+				CampaignID:     campaign3.Id,
+				Address:        addr1,
+				Shares:         share1,
+				VestingOptions: sample.ShareVestingOptions(),
+			},
+			err: types.ErrTotalShareLimit,
+		},
+		{
+			name: "create new account with shares",
+			msg: types.MsgAddMainnetVestingAccount{
+				Coordinator:    coordAddr1,
+				CampaignID:     campaign1.Id,
+				Address:        addr1,
+				Shares:         share2,
+				VestingOptions: sample.ShareVestingOptions(),
+			},
+		},
+		{
+			name: "update existing account shares",
+			msg: types.MsgAddMainnetVestingAccount{
+				Coordinator:    coordAddr1,
+				CampaignID:     campaign1.Id,
+				Address:        addr2,
+				Shares:         share2,
+				VestingOptions: sample.ShareVestingOptions(),
+			},
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			var (
+				accountExists bool
+				tmpAccount    types.MainnetVestingAccount
+				tmpCampaign   types.Campaign
+			)
+			if tc.err == nil {
+				var found bool
+				tmpCampaign, found = campaignKeeper.GetCampaign(sdkCtx, tc.msg.CampaignID)
+				require.True(t, found)
+
+				tmpAccount, accountExists = campaignKeeper.GetMainnetVestingAccount(
+					sdkCtx,
+					tc.msg.CampaignID,
+					tc.msg.Address,
+				)
+			}
+			_, err := campaignSrv.AddMainnetVestingAccount(ctx, &tc.msg)
+			if tc.err != nil {
+				require.ErrorIs(t, err, tc.err)
+				return
+			}
+			require.NoError(t, err)
+
+			account, found := campaignKeeper.GetMainnetVestingAccount(sdkCtx, tc.msg.CampaignID, tc.msg.Address)
+			require.True(t, found)
+
+			campaign, found := campaignKeeper.GetCampaign(sdkCtx, tc.msg.CampaignID)
+			require.True(t, found)
+
+			totalShares, err := account.GetTotalShares()
+			require.NoError(t, err)
+			tmpShare := types.IncreaseShares(tmpCampaign.AllocatedShares, totalShares)
+
+			if accountExists {
+				tmpAccShares, err := tmpAccount.GetTotalShares()
+				require.NoError(t, err)
+				tmpShare, err = types.DecreaseShares(tmpShare, tmpAccShares)
+				require.NoError(t, err)
+			}
+
+			equal := types.IsEqualShares(campaign.AllocatedShares, tmpShare)
+			require.True(t, equal)
```

Comment:
can we also compare the whole `VestingOptions` structure of the account equals the one given from the message